`java.io.IOException: No space left on device` within Auctionmark Benchmark

Seen this on the 24hour runs - some notes on the config where this happens:
- Each node has a 50GiB persistent volume claim attached.
- One node failing with the error - running out of space within the PvC.
- The nodes are configured with a "maxdiskcachepercentage" of 75% - so seemingly we must go over that by over 12.5GiB.

## Investigation

```
java.util.concurrent.ExecutionException: clojure.lang.ExceptionInfo: Ingestion stopped: java.util.concurrent.ExecutionException: java.io.IOException: No space left on device {}
    at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
...
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: clojure.lang.ExceptionInfo: Ingestion stopped: java.util.concurrent.ExecutionException: java.io.IOException: No space left on device
    at xtdb.await$__GT_ingester_ex.invokeStatic(await.clj:17)
    at xtdb.await$__GT_ingester_ex.invoke(await.clj:16)
    at xtdb.await$await_tx_async$fn__21938.invoke(await.clj:27)
    at xtdb.await$await_tx_async.invokeStatic(await.clj:20)
    at xtdb.await$await_tx_async.invoke(await.clj:19)
    at xtdb.indexer.Indexer.awaitTx(indexer.clj:749)
    ... 30 common frames omitted
```

# Notes / Graphs

## Failing Run

**Here's a view of the diskcache usage from the run where the node died with the above error:**
![image](https://github.com/user-attachments/assets/fb2bc31d-2458-4044-a8a2-a7e368da3187)

Sure enough, the cache was pretty full when the no disk space error hit, though don't see any immediate problems here unless there was 10gigs of other "stuff" hanging around.

## Current Run

Decided to look at the CURRENT run we have going, with a 100GiB persisent volume claim instead.

**Here's some views over the current diskcache for node 1:**
![image](https://github.com/user-attachments/assets/464a4edc-5c8c-4f4c-953c-dec1919db3c9)

As we can see:
- All stacks up to 75GiB (as expected with a volume claim of 100GiB and 75% max disk cache percentage)
- Current "in use" bytes of that should be pinned bytes + evictable bytes
- This currently totals around 44GiB.

Keeping the above in mind, decided to run `df -h` against that node, and saw the following:
```
Filesystem      Size  Used Avail Use% Mounted on
...
/dev/sde         98G   59G   40G  60% /var/lib/xtdb/buffers
...
```

59GiB! There's a gap between that and the current "in use" bytes of the bufferpool by about 15GiB. Looking at the disk usage of the `buffer` file directory:
```
15G	/var/lib/xtdb/buffers/disk-cache-1/.tmp
104K	/var/lib/xtdb/buffers/disk-cache-1/chunk-metadata
45G	/var/lib/xtdb/buffers/disk-cache-1/tables
59G	/var/lib/xtdb/buffers/disk-cache-1
```

Sure enough, 15GiB of usage in the `.tmp` folder! That WOULD be enough to push our original node over the edge, ie:
- 37.5GiB buffer pool (at total max capacity)
- 15GiB temp directory.
=
52.5GiB total size! Would exceed the total size of the persisent volume claim. 

## Summarized Problem

- We have around 15GiB of temp currently, which in the 50GiB persistent volume case would have pushed us into "run out of disk space" territory.
- All of the 15GiB of current temp are "upload<num>.arrow" files, created by the disk cache.
- Usual behaviour of "open arrow writer":
  - Call to upload-arrow-file on the object store ->
  - We write a file notification to the file notification topic ->
  - We call to .put on the diskcache, with the key and tmp-path.
    - put on the diskcache is responsible for doing the atomic move from the tmp-path to the disk cache directory.
  - So only put on the diskcache is going to clear up the tmp-path made by open-arrow-writer.

Now, we've seen a recurring info/warn of `15:47:29.039 [xtdb.compactor-558] INFO  xtdb.azure.object-store - Blob already exists for x - aborting multipart upload` type errors - Ie, when another node has already uploaded a file:
- In THAT case, we abort the multipart and throw an error in upload-arrow-file,
- Ie, we wouldnt call .put on diskcache - wouldnt make sense to!  

In that case - I don't think anything clears up the temp file, indeed, looking at the most recent temp file that has stuck around:
```
15:47 upload11723359101008666313.arrow
```

This being so - should be able to write a test against this and fix it fairly easily.

--- a/core/src/main/clojure/xtdb/buffer_pool.clj
+++ b/core/src/main/clojure/xtdb/buffer_pool.clj
@@ -159,7 +159,7 @@
   (objectSize [_ k] (Files/size (.resolve disk-store k)))
 
   (openArrowWriter [_ k rel]
-    (let [tmp-path (create-tmp-path disk-store)]
+    (let [^Path tmp-path (create-tmp-path disk-store)]
       (util/with-close-on-catch [file-ch (util/->file-channel tmp-path util/write-truncate-open-opts)
                                  unl (.startUnload rel file-ch)]
         (reify ArrowWriter
@@ -180,7 +180,9 @@
           (close [_]
             (util/close unl)
             (when (.isOpen file-ch)
-              (.close file-ch)))))))
+              (.close file-ch))
+            ;; If tmp path still present/hasn't been atomic moved, delete it on close
+            (Files/deleteIfExists tmp-path))))))
 
   EvictBufferTest
   (evict-cached-buffer! [_ k]
@@ -317,7 +319,7 @@
     (.get !os-files k))
 
   (openArrowWriter [_ k rel]
-    (let [tmp-path (.createTempPath disk-cache)]
+    (let [^Path tmp-path (.createTempPath disk-cache)]
       (util/with-close-on-catch [file-ch (util/->file-channel tmp-path util/write-truncate-open-opts)
                                  unl (.startUnload rel file-ch)]
 
@@ -337,7 +339,9 @@
           (close [_]
             (util/close unl)
             (when (.isOpen file-ch)
-              (.close file-ch)))))))
+              (.close file-ch))
+            ;; If tmp path still present/hasn't been atomic moved, delete it on close
+            (Files/deleteIfExists tmp-path))))))
 
   (putObject [_ k buffer]
     @(-> (.putObject object-store k buffer)

--- a/src/test/clojure/xtdb/buffer_pool_test.clj
+++ b/src/test/clojure/xtdb/buffer_pool_test.clj
@@ -329,3 +329,41 @@
   (tu/with-tmp-dirs #{tmp-dir}
     (with-open [bp (bp/open-local-storage tu/*allocator* (Storage/localStorage tmp-dir) (SimpleMeterRegistry.))]
       (test-list-objects bp))))
+
+(defn count-tmp-files [^Path tmp-dir]
+  (count (.listFiles (.toFile tmp-dir))))
+
+(t/deftest buffer-pool-clears-up-arrow-writer-temp-files
+  (with-open [bp (remote-test-buffer-pool)]
+    (let [^Path root-path (.getRootPath ^DiskCache (:disk-cache bp))
+          ^Path tmp-dir (.resolve root-path ".tmp")
+          schema (Schema. [(types/col-type->field "a" :i32)])]
+      (t/testing "successful uploads"
+        (with-open [rel (Relation. tu/*allocator* schema)
+                    w (.openArrowWriter bp (util/->path "aw") rel)]
+          ;; Arrow Writer Opened
+          (t/is (= 1 (count-tmp-files tmp-dir)) "temp file present")
+
+          ;; Write arrow file out
+          (let [v (.get rel "a")]
+            (.writeInt v 1)
+            (.writeBatch w)
+            (.end w))
+          
+          (t/is (= 0 (count-tmp-files tmp-dir)) "temp file removed")))
+
+      (t/testing "failing/erroring uploads"
+        (with-redefs [util/->mmap-path (fn [_] (throw (Exception. "Example error in upload arrow file")))]
+          (with-open [rel (Relation. tu/*allocator* schema)
+                      w (.openArrowWriter bp (util/->path "aw2") rel)]
+           ;; Arrow Writer Opened
+            (t/is (= 1 (count-tmp-files tmp-dir)) "temp file present")
+
+           ;; Write arrow file out, should error
+            (t/is (thrown-with-msg? Exception #"Example error in upload arrow file"
+                                    (let [v (.get rel "a")]
+                                      (.writeInt v 1)
+                                      (.writeBatch w)
+                                      (.end w)))) 
+            (t/is (= 1 (count-tmp-files tmp-dir)) "temp file still present")))
+        (t/is (= 0 (count-tmp-files tmp-dir)) "temp file removed after writer closed")))))

--- a/src/test/clojure/xtdb/buffer_pool_test.clj
+++ b/src/test/clojure/xtdb/buffer_pool_test.clj
@@ -354,16 +354,17 @@
 
       (t/testing "failing/erroring uploads"
         (with-redefs [util/->mmap-path (fn [_] (throw (Exception. "Example error in upload arrow file")))]
-          (with-open [rel (Relation. tu/*allocator* schema)
-                      w (.openArrowWriter bp (util/->path "aw2") rel)]
-           ;; Arrow Writer Opened
-            (t/is (= 1 (count-tmp-files tmp-dir)) "temp file present")
-
-           ;; Write arrow file out, should error
+          (with-open [rel (Relation. tu/*allocator* schema)]
+            ;; Using `with-close-on-catch`, as we do in trie.clj where we use the arrow writer
             (t/is (thrown-with-msg? Exception #"Example error in upload arrow file"
-                                    (let [v (.get rel "a")]
-                                      (.writeInt v 1)
-                                      (.writeBatch w)
-                                      (.end w)))) 
-            (t/is (= 1 (count-tmp-files tmp-dir)) "temp file still present")))
-        (t/is (= 0 (count-tmp-files tmp-dir)) "temp file removed after writer closed")))))
+                                    (util/with-close-on-catch [w (.openArrowWriter bp (util/->path "aw2") rel)]
+                                      ;; Arrow Writer Opened
+                                      (t/is (= 1 (count-tmp-files tmp-dir)) "temp file present")
+
+                                     ;; Write arrow file out, should error
+                                      (let [v (.get rel "a")]
+                                        (.writeInt v 1)
+                                        (.writeBatch w)
+                                        (.end w)))))
+            
+            (t/is (= 0 (count-tmp-files tmp-dir)) "temp file removed after writer errored/closed")))))))

